dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/CompilationVisitorArm.java

```diff
@@ -0,0 +1,349 @@
+import static com.dat3m.dartagnan.program.event.EventFactory.newRMWStoreExclusive;
+import static com.dat3m.dartagnan.program.event.EventFactory.newStore;
+import static com.dat3m.dartagnan.program.event.Tag.STRONG;
+import static com.dat3m.dartagnan.program.processing.Compilation.commonVisitLock;
+import static com.dat3m.dartagnan.program.processing.Compilation.commonVisitUnlock;
+
+@Options
+public class CompilationVisitorArm implements EventVisitor<List<Event>> {
+
+	@Override
+	public List<Event> visitEvent(Event e) {
+		return Collections.singletonList(e);
+	};
+
+	@Override
+	public List<Event> visitCreate(Create e) {
+        Store store = newStore(e.getAddress(), e.getMemValue(), e.getMo(), e.getCLine());
+        store.addFilters(C11.PTHREAD);
+
+        return eventSequence(
+        		AArch64.DMB.newISHBarrier(),
+                store,
+                AArch64.DMB.newISHBarrier()
+        );
+	}
+
+	@Override
+	public List<Event> visitEnd(End e) {
+        return eventSequence(
+        		AArch64.DMB.newISHBarrier(),
+        		newStore(e.getAddress(), IConst.ZERO, e.getMo()),
+                AArch64.DMB.newISHBarrier()
+        );
+	}
+
+	@Override
+	public List<Event> visitInitLock(InitLock e) {
+		return eventSequence(
+                newStore(e.getAddress(), e.getMemValue(), e.getMo())
+        );
+	}
+
+	@Override
+	public List<Event> visitJoin(Join e) {
+        List<Event> events = new ArrayList<>();
+        Register resultRegister = e.getResultRegister();
+		Load load = newLoad(resultRegister, e.getAddress(), e.getMo());
+        load.addFilters(C11.PTHREAD);
+        events.add(load);
+        events.add(AArch64.DMB.newISHBarrier());
+        events.add(newJumpUnless(new Atom(resultRegister, EQ, IConst.ZERO), e.getLabel()));
+        
+        return events;
+	}
+
+	@Override
+	public List<Event> visitLock(Lock e) {
+		return commonVisitLock(e);
+	}
+
+	@Override
+	public List<Event> visitStart(Start e) {
+        List<Event> events = new ArrayList<>();
+        Register resultRegister = e.getResultRegister();
+        events.add(newLoad(resultRegister, e.getAddress(), e.getMo()));
+        events.add(AArch64.DMB.newISHBarrier());
+        events.add(newJumpUnless(new Atom(resultRegister, EQ, IConst.ONE), e.getLabel()));
+        
+        return events;
+	}
+
+	@Override
+	public List<Event> visitUnlock(Unlock e) {
+		return commonVisitUnlock(e);
+	}
+
+	@Override
+	public List<Event> visitStoreExclusive(StoreExclusive e) {
+        RMWStoreExclusive store = newRMWStoreExclusive(e.getAddress(), e.getMemValue(), e.getMo());
+        
+        return eventSequence(
+                store,
+                newExecutionStatus(e.getResultRegister(), store)
+        );
+	}
+
+	@Override
+	public List<Event> visitRMWAbstract(RMWAbstract e) {
+		throw new IllegalArgumentException("Compilation to " + ARM8 + " is not supported for " + e.getClass().getName());
```

Comment:
So you propose to have the interface, a base implementation handling event, condjump, locks events and throwing exceptions for the one that can be compiled from a single arch?

Reply:
Yes, the base implementation would handle all events that it can handle uniformly, which is already worth it. And as default implementation for all non-uniformly compilable events, it will throw an exception.
The subclasses would then only need to implement the methods it can compile.
I think this can avoid a lot of code repetition.
Also, I wouldn't count the interface into any of this, because the interface is a generic interface for all possible visitors, not just for compilation. That's why I'm against any non-trivial default behaviour inside the interface.